src/debug.c

```diff
@@ -14,20 +23,104 @@ void gssntlm_debug_init(void)
 {
     char *env;
 
+    pthread_mutex_lock(&debug_mutex);
+
     env = secure_getenv("GSSNTLMSSP_DEBUG");
     if (env) {
         debug_fd = fopen(env, "a");
         if (debug_fd) gssntlm_debug_enabled = true;
     }
     gssntlm_debug_initialized = true;
+
+    pthread_mutex_unlock(&debug_mutex);
 }
 
 void gssntlm_debug_printf(const char *fmt, ...)
 {
     va_list ap;
 
-    va_start(ap, fmt);
-    vfprintf(debug_fd, fmt, ap);
-    va_end(ap);
-    fflush(debug_fd);
+    pthread_mutex_lock(&debug_mutex);
```

Comment:
Using a mutex per each call to gssntlm_debug_printf is a performance regression...
I understand that you need to avoid the race between changes of gssntlm_debug_initialized and debug_fd, but the price for that is too high as for me.

Can't we use ONLY debug_fd variable in debug-related code and assume that debug_fd == NULL is the indication of gssntlm_debug_enabled? In that case a pointer always fits into machine word and is always updated atomically.
Also to avoid changes of debug_fd between calls to vfprintf and fflush we can copy the value of debug_fd into local on-stack variable. Yes, I know that theoretically we still can have a race condition when old debug_fd is already closed on the moment of calling vfprintf/fflush (because gssntlm_debug_disable was called just before) BUT in the worst-case we just have the error for vfprintf/fflush and we ignore it (operation with already closed fd).

Reply:
I just recognized that debug_fd is a pointer (not int fd) so when file is closed, we can't access this memory anymore even if pointer itself is copied to local variable. Probably we can work-around it by setting debug_fd to NULL before actually calling fclose for it. However if debug_fd was changed few times (like using disabling + enabling) we probably can't avoid the race completely... So probably this way is not robust enough...

Reply:
The mutex kicks in only during debugging, so I thought it would be ok to trade a bit of performance in this case.
Can you confirm that you are still concerned about performance while debugging ?

Reply:
Note that passing NULL to vfprintf() will cause the process to crash, so I cannot just use debug_fd as an atomic, I will definitely have to copy it ... but then we can still have a crash if we point to a freed pointer... perhaps I should convert this code to use a file descriptor instead, as trying to operate on a bad fd will indeed simply cause an error and no crashes ...

Reply:
We were going to always use gssntlmssp debugging feature in our product - to be able to troubleshoot issues when they happened. I agree that it's possible to implement on-demand debugging (debugging is disabled by default but can be temporary enabled manually), but issues are often hard to reproduce so this way of troubleshooting is not robust.

Reply:
Yes, I reviewed new code with  numerical debug_fd descriptor and no mutex in gssntlm_debug_printf - this is exactly what I wanted to get! No mutex overhead in the main debug flow and no pointers make it safe, simple and robust. Thank you very much!